### ISSUE TITLE
task7: Create an lstm classification model

### ISSUE THREAD

**Issue by nlauchande** [omdena/ptsd#11]

**Comment by nlauchande** [comment]
**Target definition** :
    if sum(criterion_collumn)>=1 then criterion_labelled

Major blocker at the moment :
      - Availability training the data, architecture setup.

**Comment by kulsoom-abdullah** [comment]
I would be interested in helping with the architecture setup (if it is needed).

**Comment by albertlai431** [comment]
I've been looking at this google collab notebook: https://colab.research.google.com/drive/18skdi4ky3K-USpQuqRFGsO7ZH7XYISCo 
It works on creating a positive or negative classification of movie reviews from being trained on a large dataset. I believe it could be applied to our application as well, but the training data format would be very different.

**Comment by sylvia31096** [comment]
Owing to the fact we have a small dataset. I had a very crazy idea but I'm thinking right now we have to try anything. 
So I was thinking, what if we trained an LSTM with the descriptions of each of symptom/criterion (because what I was working on was A1 let's have that in mind) and then predict on the dataset from Colton as test and validation.

When I say description I mean, like for A1 the criterion is exposure to traumatic events, so basically a description of a traumatic event. We could scrape the web for such data and get hundreds of rows (would be tedious but possible). We would obviously need a psychologist or an expert in the field to guide us with what we should be looking at as we collect the data.

I created a colab notebook to just sort of demonstrate this. I built an LSTM model with a pretrained BERT model. 

It's predictions are not perfect but it could be due to the fact the dataset is extremely small, hyperparameters have not been tuned, the idea is downright insane or all the above.

In the notebook my created "dataset of symptoms" is not exactly how I would want it to be so I will work on that. And my code is not well commented so apologies for that also.

Here is the link to my notebook:
https://colab.research.google.com/drive/1vwbUlcZy9-3Nou3Jp0w2sGsI7gnbi5PL

**Comment by BottleSpink** [comment]
This is a great idea Sylvia - to explore criterion separately. Although, we can see what Alexis has to say about it. As you are trying BERT, I have been trying Elmo. Initially, I tried word2vec but for getting contextualized embedding - Bert, Elmo will be the better options. For now, I have used Shikha's LSTM pipeline as such. My next step is to try bi-LSTM. I will try to upload my results from Colab by tomorrow.

**Comment by rrupam** [comment]
Hey, I've trained a model on rotten tomatoes movie reviews dataset. I've used a pretrained bert model and retrained it on the reviews dataset. It has 5 classes. The model is trained and saved. To test it on new data, we can follow the steps from code block 11 ('TESTING'). Here's the link to the google drive folder with all relevant files : https://drive.google.com/drive/folders/1tyBUHSccyDP_xKaLmI9G5MDqSkOvCsAl

**Comment by rrupam** [comment]
An update to the above. I've trained the BERT model using Colton's dataset with Alexis's annotation. Here's the code: https://colab.research.google.com/drive/1kMMpG8ObcJxAf74oN7uS1Zil3S_FZqrs
The accuracy is pretty high right now~96%, but I think it is also because most of the classes are 0. 
It is still useful to see that we can use this model and train it on any dataset, with just a spot of preprocessing.

**Comment by sylvia31096** [comment]
@rrupam Nice! Is the accuracy based on the training set? And also the text you've used, is it from both the therapist and client?

**Comment by rrupam** [comment]
@sylvia31096 Thanks Sylvia. Yes, I did a train-test split on the training set and the accuracy is based on the validation set. You're right, I didn't pick out the client text, which is probably the correct thing to do here. I will fix this. Thanks.

**Comment by MamadouBousso** [comment]
Hi, this is a link https://colab.research.google.com/drive/1vlKEJPA2AO0o0mk2M501VyFg3h5QliRo#scrollTo=9VY-M3D0fV0b for a notebook on LSTM and some data discovery based on co-occurrence matrix and word2vec. It is based on the job done in the group. The LSTM model is not doing well because of the lack of data. I have problems putting the data file with google colab. It' a little part of the first transcript done by Alexis. I will join the group of Alexis for annotating in order to get more data.
[PTSD_annotated_data1.zip](https://github.com/omdena/ptsd/files/3463225/PTSD_annotated_data1.zip)
